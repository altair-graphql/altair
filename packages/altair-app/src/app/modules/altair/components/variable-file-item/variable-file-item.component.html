<div
  class="variables-file-item"
  [ngClass]="{
    'variables-file-item--invalid': invalidFileData,
    'variables-file-item--warning': showWarning
  }"
>
  <div class="variables-file-item-name-input-wrapper">
    <span
      class="variables-file-item-switch"
      [ngClass]="{
        'variables-file-item-switch--enabled': fileVariable?.isMultiple
      }"
      [popper]="'Toggle single/multiple file mode'"
      (click)="fileVariableIsMultipleChange.emit(!fileVariable?.isMultiple)"
    >
      <span *ngIf="!fileVariable?.isMultiple">1</span>
      <span *ngIf="fileVariable?.isMultiple">*</span>
    </span>
    <input
      type="text"
      class="variables-file-item-name-input"
      data-tid="file_name"
      placeholder="{{ 'FILE_NAME_PLACEHOLDER' | translate }}"
      [ngModel]="fileVariable?.name"
      (ngModelChange)="fileVariableNameChange.emit($event)"
    />
  </div>
  <div class="variables-file-item-input-wrapper">
    <input
      #fileEl
      type="file"
      class="variables-file-item-input"
      [attr.multiple]="fileVariable?.isMultiple ? '' : null"
      (change)="onSelectFiles()"
    />
    <span class="variables-file-item-selected-total" [popper]="filesText">
      {{ validFileData.length }} file(s)
    </span>
    <button
      class="app-button variables-file-item-button"
      (click)="fileEl.click()"
    >
      <app-icon name="paperclip"></app-icon>
      {{ 'FILES_SELECT_FILES' | translate }}
    </button>
  </div>
  <div
    class="variables-file-item-input-delete"
    (click)="deleteFileVariableChange.emit()"
  >
    &#x00D7;
  </div>
</div>
