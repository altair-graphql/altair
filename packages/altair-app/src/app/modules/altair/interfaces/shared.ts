export interface IDictionary<V = any> {
  [key: string]: V;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type TODO = any;

export type UnknownError = TODO;

export interface TrackByIdItem {
  id: string;
}
