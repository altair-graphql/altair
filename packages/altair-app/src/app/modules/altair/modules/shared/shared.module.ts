import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';

import { DragDropModule } from '@angular/cdk/drag-drop';
import { TranslateModule } from '@ngx-translate/core';
import { NgPipesModule } from 'ngx-pipes';
import { NguiAutoCompleteModule } from '@ngui/auto-complete';
import { MarkdownModule } from 'ngx-markdown';
import { HotToastModule } from '@ngneat/hot-toast';
import { ResizableModule } from 'angular-resizable-element';

import { NzButtonModule } from 'ng-zorro-antd/button';
import { NzFormModule } from 'ng-zorro-antd/form';
import { NzSelectModule } from 'ng-zorro-antd/select';
import { NzModalModule } from 'ng-zorro-antd/modal';
import { NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { NzSwitchModule } from 'ng-zorro-antd/switch';
import { NzListModule } from 'ng-zorro-antd/list';
import { NzRadioModule } from 'ng-zorro-antd/radio';
import { NzAlertModule } from 'ng-zorro-antd/alert';
import { NzLayoutModule } from 'ng-zorro-antd/layout';
import { NzMenuModule } from 'ng-zorro-antd/menu';
import { NzCollapseModule } from 'ng-zorro-antd/collapse';
import { NzInputModule } from 'ng-zorro-antd/input';
import { NzInputNumberModule } from 'ng-zorro-antd/input-number';
import { NzAutocompleteModule } from 'ng-zorro-antd/auto-complete';
import { NzCheckboxModule } from 'ng-zorro-antd/checkbox';
import { NzToolTipModule } from 'ng-zorro-antd/tooltip';
import { NzTabsModule } from 'ng-zorro-antd/tabs';
import { NzTableModule } from 'ng-zorro-antd/table';
import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
import { NzGridModule } from 'ng-zorro-antd/grid';
import { NzPopoverModule } from 'ng-zorro-antd/popover';

import { IconsModule } from '../icons/icons.module';
import { NzAvatarModule } from 'ng-zorro-antd/avatar';

const AntdComponentModules = [
  NzButtonModule,
  NzInputModule,
  NzInputNumberModule,
  NzAutocompleteModule,
  NzCheckboxModule,
  NzFormModule,
  NzGridModule,
  NzTabsModule,
  NzSelectModule,
  NzModalModule,
  NzDropDownModule,
  NzSwitchModule,
  NzListModule,
  NzRadioModule,
  NzAlertModule,
  NzLayoutModule,
  NzMenuModule,
  NzCollapseModule,
  NzToolTipModule,
  NzTableModule,
  NzTreeSelectModule,
  NzPopoverModule,
  NzAvatarModule,
  // NgZorroAntdModule,
];

@NgModule({
  imports: [
    CommonModule,
    DragDropModule,
    TranslateModule.forRoot(),
    NgPipesModule,
    NguiAutoCompleteModule, // use antd
    MarkdownModule.forRoot(),
    ResizableModule,
    IconsModule,
    // ToastrModule.forRoot({
    //   newestOnTop: false,
    //   closeButton: true,
    //   positionClass: 'toast-top-center',
    //   enableHtml: true,
    //   countDuplicates: true,
    //   preventDuplicates: true,
    //   resetTimeoutOnDuplicate: true,
    // }),
    // HotToastModule.forRoot({
    //   position: 'top-center',
    //   reverseOrder: true,
    //   dismissible: true,
    //   closeStyle: {
    //     backgroundImage: `none`,
    //   },
    // }),
    ...AntdComponentModules,
  ],

  // DON'T FORGET TO EXPORT MODULE
  exports: [
    DragDropModule,
    TranslateModule,
    NgPipesModule,
    NguiAutoCompleteModule,
    MarkdownModule,
    ResizableModule,
    IconsModule,
    HotToastModule,
    ...AntdComponentModules,
  ],
})
export class SharedModule {
  static forRoot(): ModuleWithProviders<SharedModule> {
    return {
      ngModule: SharedModule,
      providers: [],
    };
  }
}
