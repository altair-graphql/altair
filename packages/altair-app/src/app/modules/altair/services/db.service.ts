import { of as observableOf, Observable, Subscriber } from 'rxjs';
import { Injectable } from '@angular/core';

@Injectable()
export class DbService {
  storagePrefix = 'altair:';
  servicePrefix = 'db:';

  /**
   * Gets the item with the exact name specified
   * @param key
   */
  getItemByExactKey(key: string): Observable<any> {
    const dbValue = localStorage.getItem(key);

    return Observable.create((observer: Subscriber<any>) => {
      if (dbValue) {
        try {
          const parsedValue = JSON.parse(dbValue);
          observer.next(parsedValue.value);
        } catch (err) {
          observer.next(dbValue);
        }
      } else {
        observer.next(null);
      }
    });
  }

  /**
   * Gets the item with the key provided
   * The key is retrieved with the application-specific key
   * @param key
   */
  getItem(key: string): Observable<any> {
    return this.getItemByExactKey(this.getItemName(key));
  }

  /**
   * Stores a value with the specified application key
   * @param key
   * @param value
   */
  setItem(key: string, value: unknown): Observable<any> {
    if (key && typeof value !== 'undefined') {
      const dbValue = {
        value,
      };
      localStorage.setItem(this.getItemName(key), JSON.stringify(dbValue));
    }

    return new Observable((obs) => obs.next(null));
  }

  /**
   * Removes an item with the specified exact key
   * @param key
   */
  removeItemByExactKey(key: string): Observable<any> {
    localStorage.removeItem(key);

    return new Observable((obs) => obs.next(null));
  }

  /**
   * Removes an item with the specified application key
   * @param key
   */
  removeItem(key: string) {
    return this.removeItemByExactKey(this.getItemName(key));
  }

  /**
   * Gets all the keys of the items stored
   */
  getAllKeys() {
    // Get all the items in the local storage that is specific to the app
    const allKeys = Object.keys(localStorage).filter((key: string) =>
      key.includes(this.storagePrefix)
    );

    // return Observable.combineLatest(allKeys.map(key => this.getItemByExactKey(key)));
    return observableOf(allKeys);
  }

  /**
   * Used to get the application-specific key
   * @param key The unique key for the data
   */
  private getItemName(key: string) {
    return `${this.storagePrefix}${this.servicePrefix}${key}`;
  }
}
