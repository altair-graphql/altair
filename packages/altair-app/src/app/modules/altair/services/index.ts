export { ApiService } from './api/api.service';
export { GqlService } from './gql/gql.service';
export { DbService } from './db.service';
export { WindowService } from './window.service';
export { NotifyService } from './notify/notify.service';
export { DonationService } from './donation.service';
export { ElectronAppService } from './electron-app/electron-app.service';
export { KeybinderService } from './keybinder/keybinder.service';
export { StorageService } from './storage/storage.service';
export { QueryCollectionService } from './query-collection/query-collection.service';
export { EnvironmentService } from './environment/environment.service';
export { PluginRegistryService } from './plugin/plugin-registry.service';
export { PluginEventService } from './plugin/plugin-event.service';
export { PreRequestService } from './pre-request/pre-request.service';
export { ThemeRegistryService } from './theme/theme-registry.service';
export { SubscriptionProviderRegistryService } from './subscriptions/subscription-provider-registry.service';
export { PluginContextService } from './plugin/context/plugin-context.service';
export { QueryService } from './query/query.service';
export { AccountService } from './account/account.service';
