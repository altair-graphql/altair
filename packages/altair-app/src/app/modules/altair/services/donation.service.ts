import { Observable, zip } from 'rxjs';
import { Injectable } from '@angular/core';

import { DbService } from './db.service';
import { uaSeedHash } from '../utils/simple_hash';
import { map } from 'rxjs/operators';
import { AltairConfig } from 'altair-graphql-core/build/config';
import { Store } from '@ngrx/store';
import { RootState } from 'altair-graphql-core/build/types/state/state.interfaces';

@Injectable()
export class DonationService {
  private actionCountKey = 'dac';
  private seedKey = 'ds';
  private hashKey = 'dh';
  private seedBuff = 100000;

  constructor(
    private dbService: DbService,
    private store: Store<RootState>,
    private altairConfig: AltairConfig
  ) {}

  donated() {
    const seed = Math.random() * this.seedBuff;

    // Store the seed
    this.dbService.setItem(this.seedKey, seed);

    // Store the seed hash
    this.dbService.setItem(this.hashKey, uaSeedHash(seed.toString()));

    // Reset the count
    this.dbService.setItem(this.actionCountKey, 0);
  }

  /**
   * counts the action and checks if the action is eligible to display the donation alert
   */
  trackAndCheckIfEligible(): Observable<boolean> {
    /**
     * Check if the count threshold has been reached.
     * Check if random seed exist
     * Check if hash of (seed+ua) matches [donated]
     * ~~ Show alert ~~
     * Reset counter
     */
    const actionCount$ = this.dbService.getItem(this.actionCountKey);
    const seed$ = this.dbService.getItem(this.seedKey);
    const curHash$ = this.dbService.getItem(this.hashKey);
    const account$ = this.store.select('account');

    return zip(actionCount$, seed$, curHash$, account$).pipe(
      map(([actionCount, seed, curHash, account]) => {
        if (account.plan?.id === 'pro') {
          // Reset count
          this.dbService.setItem(this.actionCountKey, 0);
          return false;
        }
        if (
          actionCount &&
          actionCount >= this.altairConfig.donation.action_count_threshold
        ) {
          // Reset count
          this.dbService.setItem(this.actionCountKey, 0);

          if (seed && uaSeedHash(seed) === curHash) {
            // User has donated already
            return false;
          } else {
            // User has not donated
            return true;
          }
        } else {
          // Increment count
          this.dbService.setItem(this.actionCountKey, actionCount + 1);

          return false;
        }
      })
    );
  }
}
