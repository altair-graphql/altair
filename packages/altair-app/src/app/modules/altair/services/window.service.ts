
import { EMPTY, from, Observable } from 'rxjs';

import { first, tap, map, switchMap } from 'rxjs/operators';
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';

import uuid from 'uuid/v4';

import * as fromQuery from '../store/query/query.reducer';

import * as fromRoot from '../store';

import * as queryActions from '../store/query/query.action';
import * as headerActions from '../store/headers/headers.action';
import * as variableActions from '../store/variables/variables.action';
import * as windowActions from '../store/windows/windows.action';
import * as windowsMetaActions from '../store/windows-meta/windows-meta.action';
import * as preRequestActions from '../store/pre-request/pre-request.action';
import * as postRequestActions from '../store/post-request/post-request.action';
import * as streamActions from '../store/stream/stream.action';
import * as localActions from '../store/local/local.action';
import * as gqlSchemaActions from '../store/gql-schema/gql-schema.action';
import * as layoutActions from '../store/layout/layout.action';

import { getFileStr, mapToKeyValueList } from '../utils';
import { parseCurlToObj } from '../utils/curl';
import { debug } from '../utils/logger';
import { GqlService } from './gql/gql.service';
import { RootState } from 'altair-graphql-core/build/types/state/state.interfaces';
import { ExportWindowState } from 'altair-graphql-core/build/types/state/window.interfaces';
import { QueryCollectionService } from './query-collection/query-collection.service';


interface ImportWindowDataOptions {
  fixedTitle?: boolean;
}

@Injectable()
export class WindowService {

  constructor(
    private store: Store<RootState>,
    private gqlService: GqlService,
    private collectionService: QueryCollectionService,
  ) { }

  newWindow(opts: { title?: string, url?: string, collectionId?: number, windowIdInCollection?: string, fixedTitle?: boolean } = {}) {
    return this.store.pipe(
      first(),
      map(state => {
        const url = opts.url || fromQuery.getInitialState().url || (
          state.windowsMeta.activeWindowId &&
          state.windows[state.windowsMeta.activeWindowId]?.query.url || ''
        )

        const newWindow = {
          windowId: uuid(),
          title: opts.title || `Window ${Object.keys(state.windows).length + 1}`,
          url,
          collectionId: opts.collectionId,
          windowIdInCollection: opts.windowIdInCollection,
          fixedTitle: opts.fixedTitle,
        };

        this.store.dispatch(new windowActions.AddWindowAction(newWindow));

        this.setupWindow(newWindow.windowId);

        return newWindow;
      }),
    );
  }

  removeWindow(windowId: string) {
    this.cleanupWindow(windowId);

    return this.store.pipe(
      first(),
      tap((data) => {
        const window = data.windows[windowId];
        if (window) {
          this.store.dispatch(new localActions.PushClosedWindowToLocalAction({ window }));
        }
        this.store.dispatch(new windowActions.RemoveWindowAction({ windowId }));
      }),
    );
  }

  duplicateWindow(windowId: string) {
    return this.store.pipe(first()).subscribe(data => {
      const window = data.windows[windowId];
      if (!window) {
        return;
      }
      const clonedWindow = { ...window };

      const windowData: ExportWindowState = {
        version: 1,
        type: 'window',
        query: clonedWindow.query.query || '',
        apiUrl: clonedWindow.query.url,
        variables: clonedWindow.variables.variables,
        subscriptionUrl: clonedWindow.query.subscriptionUrl,
        subscriptionConnectionParams: clonedWindow.query.subscriptionConnectionParams,
        headers: clonedWindow.headers,
        windowName: `${clonedWindow.layout.title} (Copy)`,
        preRequestScript: clonedWindow.preRequest.script,
        preRequestScriptEnabled: clonedWindow.preRequest.enabled,
        gqlSchema: clonedWindow.schema.schema,
      };

      return this.importWindowData(windowData);
    });
  }

  getWindowExportData(windowId: string) {
    return this.store.pipe(
      first(),
      map(state => {
        const window = state.windows[windowId];
        if (!window) {
          return;
        }
        const clonedWindow = { ...window };

        // TODO: Check that there is data to be exported

        return {
          version: 1 as const,
          type: 'window' as const,
          query: clonedWindow.query.query || '',
          apiUrl: clonedWindow.query.url,
          variables: clonedWindow.variables.variables,
          subscriptionUrl: clonedWindow.query.subscriptionUrl,
          subscriptionConnectionParams: clonedWindow.query.subscriptionConnectionParams,
          headers: clonedWindow.headers,
          windowName: clonedWindow.layout.title,
          preRequestScript: clonedWindow.preRequest.script,
          preRequestScriptEnabled: clonedWindow.preRequest.enabled,
          postRequestScript: clonedWindow.postRequest.script,
          postRequestScriptEnabled: clonedWindow.postRequest.enabled,
        };
      }),
    );
  }

  importWindowDataFromJson(data: string) {
    if (!data) {
      throw new Error('String is empty.');
    }

    try {
      return this.importWindowData(JSON.parse(data), { fixedTitle: true });
    } catch (err) {
      try {
        // For a period, the JSON data was URI encoded.
        // Maybe that is the problem with this data.
        debug.log('(Second attempt) Trying to decode JSON data...');
        return this.importWindowData(JSON.parse(decodeURIComponent(data)), { fixedTitle: true });
      } catch (err) {}
      debug.log('The file is invalid.', err);
    }
  }

  importWindowDataFromCurl(curlStr: string) {
    if (!curlStr) {
      throw new Error('String is empty.');
    }

    try {
      const curlObj = parseCurlToObj(curlStr);
      const windowData: ExportWindowState = {
        version: 1,
        type: 'window',
        query: curlObj.body ? JSON.parse(curlObj.body).query : '',
        apiUrl: curlObj.url,
        variables: curlObj.body ? JSON.stringify(JSON.parse(curlObj.body).variables) : '',
        subscriptionUrl: '',
        subscriptionConnectionParams: '',
        headers: curlObj.headers ? mapToKeyValueList(curlObj.headers) : [],
        windowName: `cURL-${Date.now()}`,
        preRequestScript: '',
        preRequestScriptEnabled: false,
      };

      return this.importWindowData(windowData);
    } catch (err) {
      debug.log('The file is invalid.', err);
    }
  }

  /**
   * Import the window represented by the provided data string
   * @param data window data string
   */
  importWindowData(data: ExportWindowState, options: ImportWindowDataOptions = {}) {
    try {
      // Verify file's content
      if (!data) {
        throw new Error('Object is empty.');
      }
      if (!data.version || !data.type || data.type !== 'window') {
        throw new Error('File is not a valid Altair query file.');
      }

      // Importing window data...
      // Add new window
      // Set window name
      // Set API URL
      // Set query
      // Set headers
      // Set variables
      // Set subscription URL
      this.newWindow({
        title: data.windowName,
        url: data.apiUrl,
        collectionId: data.collectionId,
        windowIdInCollection: data.windowIdInCollection,
        fixedTitle: options.fixedTitle,
      }).subscribe(newWindow => {
        const windowId = newWindow.windowId;

        if (data.apiUrl) {
          this.store.dispatch(new queryActions.SetUrlAction({ url: data.apiUrl }, windowId));
          this.store.dispatch(new queryActions.SendIntrospectionQueryRequestAction(windowId));
        }

        if (data.query) {
          this.store.dispatch(new queryActions.SetQueryAction(data.query, windowId));
        }

        if (data.headers.length) {
          this.store.dispatch(new headerActions.SetHeadersAction({ headers: data.headers }, windowId));
        }

        if (data.variables) {
          this.store.dispatch(new variableActions.UpdateVariablesAction(data.variables, windowId));
        }

        if (data.subscriptionUrl) {
          this.store.dispatch(new queryActions.SetSubscriptionUrlAction({ subscriptionUrl: data.subscriptionUrl }, windowId));
        }

        if (data.subscriptionConnectionParams) {
          this.store.dispatch(
            new queryActions.SetSubscriptionConnectionParamsAction(windowId, { connectionParams: data.subscriptionConnectionParams })
          );
        }

        if (data.subscriptionProvider) {
          this.store.dispatch(new queryActions.SetSubscriptionProviderIdAction(windowId, { providerId: data.subscriptionProvider }));
        }

        if (data.preRequestScriptEnabled) {
          this.store.dispatch(new preRequestActions.SetPreRequestEnabledAction(windowId, { enabled: data.preRequestScriptEnabled }));
        }
        if (data.preRequestScript) {
          this.store.dispatch(new preRequestActions.SetPreRequestScriptAction(windowId, { script: data.preRequestScript }));
        }

        if (data.postRequestScriptEnabled) {
          this.store.dispatch(new postRequestActions.SetPostRequestEnabledAction(windowId, { enabled: data.postRequestScriptEnabled }));
        }
        if (data.postRequestScript) {
          this.store.dispatch(new postRequestActions.SetPostRequestScriptAction(windowId, { script: data.postRequestScript }));
        }

        if (data.gqlSchema) {
          this.store.dispatch(new gqlSchemaActions.SetSchemaAction(windowId, data.gqlSchema));
        }

        this.store.dispatch(new windowsMetaActions.SetActiveWindowIdAction({ windowId }));
      });
    } catch (err) {
      debug.log('Something went wrong while importing the data.', err);
    }
  }

  /**
   * Parse data and import as identified type
   * @param dataStr data
   */
  importStringData(dataStr: string) {
    const invalidFileError = new Error('Invalid Altair window file.');
    const emptyWindowData: ExportWindowState = {
      version: 1,
      type: 'window',
      apiUrl: '',
      headers: [],
      preRequestScript: '',
      preRequestScriptEnabled: false,
      query: '',
      subscriptionUrl: '',
      subscriptionConnectionParams: '',
      variables: '{}',
      windowName: '',
    };

    try {
      let parsed: ExportWindowState;
      try {
        parsed = JSON.parse(dataStr);
      } catch (err) {
        parsed = JSON.parse(decodeURIComponent(dataStr));
      }

      if (parsed.type === 'window' && parsed.version === 1) {
        return this.importWindowData(parsed, { fixedTitle: true });
      }
      throw invalidFileError;
    } catch (err) {
      debug.log('Invalid Altair window file.', err);
      try {
        // Check if sdl file
        const schema = this.gqlService.sdlToSchema(dataStr);
        if (schema) {
          // Import only schema
          return this.importWindowData({
            ...emptyWindowData,
            version: 1,
            type: 'window',
            gqlSchema: schema,
          });
        }
        throw invalidFileError;
      } catch (sdlError) {
        debug.log('Invalid SDL file.', sdlError);
        try {
          // Else check if graphql query
          const operations = this.gqlService.getOperations(dataStr);
          debug.log(operations);
          if (operations && operations.length) {
            // Import only query
            return this.importWindowData({
              ...emptyWindowData,
              version: 1,
              type: 'window',
              query: dataStr,
            });
          }
          throw invalidFileError;
        } catch (queryError) {
          throw queryError;
        }
      }
    }
  }

  /**
   * Handle the imported file
   */
  async handleImportedFile(files: FileList) {
    const dataStr = await getFileStr(files);
    try {
      this.importStringData(dataStr);
    } catch (error) {
      debug.log('There was an issue importing the file.', error);
    }
  }

  /**
   * Carry out any necessary house cleaning tasks.
   */
  setupWindow(windowId: string) {
    // Validate that query is ACTUALLY in an existing collection
    this.getWindowState(windowId).pipe(
      first(),
      switchMap(data => {
        if (data?.layout.collectionId && data?.layout.windowIdInCollection) {
          return from(this.collectionService.getCollectionByID(data.layout.collectionId)).pipe(
            map(collection => {
              if (collection) {
                const query = collection.queries.find(q => q.id === data.layout.windowIdInCollection);
                return !!query;
              }

              return false;
            }),
          );
        }

        return EMPTY;
      }),
    ).subscribe(isValidCollectionQuery => {
      if (!isValidCollectionQuery) {
        this.store.dispatch(new layoutActions.SetWindowIdInCollectionAction(windowId, {}));
      }
    });

    this.store.dispatch(new queryActions.SetSubscriptionResponseListAction(windowId, { list: [] }));
    this.store.dispatch(new queryActions.StopSubscriptionAction(windowId));
    this.store.dispatch(new streamActions.StopStreamClientAction(windowId));
    this.store.dispatch(new streamActions.StartStreamClientAction(windowId));

    this.store.pipe(
      first(),
      map(data => data.settings['schema.reloadOnStart']),
    ).subscribe(shouldReloadSchema => {
      if (shouldReloadSchema) {
        this.store.dispatch(new queryActions.SendIntrospectionQueryRequestAction(windowId));
      }
    });
  }

  private getWindowState(windowId: string) {
    return this.store.pipe(select(fromRoot.selectWindowState(windowId)));
  }
  private cleanupWindow(windowId: string) {
    this.store.dispatch(new queryActions.StopSubscriptionAction(windowId));
    this.store.dispatch(new streamActions.StopStreamClientAction(windowId));
  }
}
