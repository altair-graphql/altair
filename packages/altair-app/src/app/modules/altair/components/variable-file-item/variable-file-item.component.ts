import {
  Component,
  OnInit,
  Input,
  Output,
  EventEmitter,
  ViewChild,
  ElementRef,
  OnChanges,
  SimpleChanges,
} from '@angular/core';
import { FileVariable } from 'altair-graphql-core/build/types/state/variable.interfaces';
import { StorageService } from '../../services';
import * as fromVariables from '../../store/variables/variables.reducer';
import { truncateText } from '../../utils';

@Component({
  selector: 'app-variable-file-item',
  templateUrl: './variable-file-item.component.html',
  styles: [],
})
export class VariableFileItemComponent implements OnInit, OnChanges {
  @Input() fileVariable?: FileVariable;

  @Output() fileVariableNameChange = new EventEmitter();
  @Output() fileVariableDataChange = new EventEmitter<{
    files: File[];
    fromCache?: boolean;
  }>();
  @Output() fileVariableIsMultipleChange = new EventEmitter();
  @Output() deleteFileVariableChange = new EventEmitter();

  @ViewChild('fileEl', { static: true }) fileEl?: ElementRef<HTMLInputElement>;

  validFileData: File[] = [];
  invalidFileData = false;
  showWarning = false;
  filesText = '';

  constructor(private storageService: StorageService) {}

  async ngOnInit() {
    if (this.fileVariable) {
      this.updateLocalState(this.fileVariable);
      if (this.invalidFileData) {
        if (this.fileVariable.id) {
          const selectedFiles = await this.storageService.selectedFiles.get(
            this.fileVariable.id
          );

          if (selectedFiles) {
            return this.fileVariableDataChange.emit({
              files: Array.from(selectedFiles.data),
              fromCache: true,
            });
          }
        }
      }
    }
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes?.fileVariable?.currentValue?.data?.length) {
      this.updateLocalState(changes.fileVariable.currentValue);
    }
  }

  onSelectFiles() {
    if (!this.fileEl) {
      return;
    }

    const files = this.fileEl.nativeElement.files;
    if (!files) {
      return;
    }

    return this.fileVariableDataChange.emit({ files: Array.from(files) });
  }

  isValidFileData(fileVariable: FileVariable) {}

  updateLocalState(fileVariable: FileVariable) {
    this.validFileData = Array.isArray(fileVariable.data)
      ? fileVariable.data.filter((data) => data instanceof File)
      : [];
    this.invalidFileData =
      (this.fileVariable?.data as [])?.length > this.validFileData.length;
    this.showWarning = Boolean(
      !fileVariable?.isMultiple && (fileVariable.data as [])?.length > 1
    );

    if (this.invalidFileData) {
      this.filesText = 'Invalid file data. Select the file again.';
    } else if (this.showWarning) {
      this.filesText = 'In single mode, only the first file will be used.';
    } else if (this.validFileData.length) {
      this.filesText = this.validFileData.map((data) => data.name).join(', ');
    } else {
      this.filesText = 'No files selected';
    }
    this.filesText = truncateText(this.filesText);
  }
}
