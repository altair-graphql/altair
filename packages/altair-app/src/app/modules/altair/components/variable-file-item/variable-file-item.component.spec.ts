import { describe, expect, it } from '@jest/globals';

import { VariableFileItemComponent } from './variable-file-item.component';
import { NgxTestWrapper, mount, mock } from '../../../../../testing';
import { MockModule } from 'ng-mocks';
import { SharedModule } from '../../modules/shared/shared.module';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { StorageService } from '../../services';

describe('VariableFileItemComponent', () => {
  let wrapper: NgxTestWrapper<VariableFileItemComponent>;

  beforeEach(async () => {
    wrapper = await mount({
      component: VariableFileItemComponent,
      imports: [MockModule(SharedModule)],
      providers: [
        {
          provide: StorageService,
          useValue: mock(),
        },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should create', () => {
    expect(wrapper.componentInstance).toBeTruthy();
  });

  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
