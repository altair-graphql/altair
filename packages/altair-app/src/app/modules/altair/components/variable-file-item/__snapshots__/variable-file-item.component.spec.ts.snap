// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariableFileItemComponent should render correctly 1`] = `
<app-variable-file-item>
  <div
    class="variables-file-item"
  >
    <div
      class="variables-file-item-name-input-wrapper"
    >
      <span
        class="variables-file-item-switch"
        nz-tooltip=""
      >
        <span>
          1
        </span>
        
        
      </span>
      
      <input
        class="variables-file-item-name-input"
        data-tid="file_name"
        placeholder=""
        type="text"
      />
      
    </div>
    <div
      class="variables-file-item-input-wrapper"
    >
      <input
        class="variables-file-item-input"
        type="file"
      />
      <span
        class="variables-file-item-selected-total"
        nz-tooltip=""
      >
         0 file(s) 
      </span>
      
      <button
        class="app-button variables-file-item-button"
      >
        <app-icon
          name="paperclip"
        />
          
      </button>
    </div>
    <div
      class="variables-file-item-input-delete"
    >
       Ã— 
    </div>
  </div>
</app-variable-file-item>
`;
